numba/parfor.py

```diff
@@ -0,0 +1,1246 @@
+    """
+    last_label = max(parfor.loop_body.keys())
+    reductions = {}
+    names = []
+    parfor_params = get_parfor_params(parfor)
+    for blk in parfor.loop_body.values():
+        for stmt in blk.body:
+            if isinstance(stmt, ir.Assign) and isinstance(stmt.value, ir.Expr) and stmt.value.op == "inplace_binop":
+                name = stmt.value.lhs.name
+                if name in parfor_params:
+                    names.append(name)
+                    reductions[name] = (stmt.value.fn, stmt.value.immutable_fn)
+    return sorted(names), reductions
+
+def visit_vars_parfor(parfor, callback, cbdata):
+    if config.DEBUG_ARRAY_OPT==1:
+        print("visiting parfor vars for:",parfor)
+        print("cbdata: ", sorted(cbdata.items()))
+    for l in parfor.loop_nests:
+        l.index_variable = visit_vars_inner(l.index_variable, callback, cbdata)
+        l.range_variable = visit_vars_inner(l.range_variable, callback, cbdata)
+    visit_vars({-1:parfor.init_block}, callback, cbdata)
+    visit_vars(parfor.loop_body, callback, cbdata)
+    return
+
+# add call to visit parfor variable
+ir_utils.visit_vars_extensions[Parfor] = visit_vars_parfor
+
+def parfor_defs(parfor):
+    """list variables written in this parfor by recursively
+    calling compute_use_defs() on body and combining block defs.
+    """
+    all_defs = set()
+    # index variables are sematically defined here
+    for l in parfor.loop_nests:
+        all_defs.add(l.index_variable.name)
+
+    # all defs of body blocks
+    for l,b in parfor.loop_body.items():
+        for stmt in b.body:
+            if isinstance(stmt, ir.Assign):
+                all_defs.add(stmt.target.name)
+            elif isinstance(stmt, Parfor):
+                all_defs.update(parfor_defs(stmt))
+
+    # all defs of init block
+    for stmt in parfor.init_block.body:
+        if isinstance(stmt, ir.Assign):
+            all_defs.add(stmt.target.name)
+        elif isinstance(stmt, Parfor):
+            all_defs.update(parfor_defs(stmt))
+
+    return all_defs
+
+analysis.ir_extension_defs[Parfor] = parfor_defs
+
+def parfor_insert_dels(parfor, curr_dead_set):
+    """insert dels in parfor. input: dead variable set right after parfor.
+    returns the variables for which del was inserted.
+    """
+    blocks = wrap_parfor_blocks(parfor)
+    cfg = compute_cfg_from_blocks(blocks)
+    usedefs = compute_use_defs(blocks)
+    live_map = compute_live_map(cfg, blocks, usedefs.usemap, usedefs.defmap)
+    dead_map = compute_dead_maps(cfg, blocks, live_map, usedefs.defmap)
+    # treat loop variables and size variables as live
+    loop_vars = {l.range_variable.name for l in parfor.loop_nests}
+    loop_vars |= {l.index_variable.name for l in parfor.loop_nests}
+    for var_list in parfor.array_analysis.array_size_vars.values():
+        loop_vars |= {v.name for v in var_list if isinstance(v, ir.Var)}
+    dead_set = set()
+    # TODO: handle escaping deads
```

Comment:
Is the escaping deads handled already?  If not, can you explain what it is and the consequences of not handling them?